### PR TITLE
Ability to store & remove server owner & deployment entries in dropdowns

### PR DESCRIPTION
**Pull request by markgrahamdawson**
Closes https://github.com/cylc/cylc-ui/issues/1457

- Added functionality for storing server owner and deployment in local storage.
- The state of the reactive array for deployments or owners is copied to a local storage variable each time it is changed.
- User can also remove values that have been stored.
- User cannot remove the initial values `onLoad` values.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by markgrahamdawson** [comment]
When I added the 'server owner' and 'deployment' dropdowns a few weeks back I [initially added some tests](https://github.com/cylc/cylc-ui/pull/1444/commits/20c435c4299d854898536d4a80a375b34423d216) but then went back and removed them (they didnt get merged in). I cant remember why the tests were seen as not required. 
Should I try and add some end to end tests here @MetRonnie ?

**Comment by markgrahamdawson** [comment]
There is a fair bit of duplicate code as the server owner combo box is doing the same thing as the deployment combo box - I could make a separate combobox component which can be used for both. However it doesnt look like that level of componentization would be consistent with rest of app.
@MetRonnie

**Comment by markgrahamdawson** [comment]
Im not sure this requires a changes.md entry as its really just finishing off a feature that's already implemented in #1444

**Comment by MetRonnie** [comment]
> When I added the 'server owner' and 'deployment' dropdowns a few weeks back I [initially added some tests](https://github.com/cylc/cylc-ui/pull/1444/commits/20c435c4299d854898536d4a80a375b34423d216) but then went back and removed them (they didnt get merged in). I cant remember why the tests were seen as not required. Should I try and add some end to end tests here @MetRonnie ?

I think those tests were moved from `dashboard.cy.js` to `header.cy.js`, not removed?

**Comment by markgrahamdawson** [comment]
TODO: I could add a test to header.cy.js to check if the combobox content remains after a a page refresh - or thats its been saved to local storage if there is a good way of accessing local storage with cypress...
https://docs.cypress.io/api/commands/getalllocalstorage

File: src/components/cylc/Header.vue [hunk 1]
```diff
@@ -47,26 +47,70 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+                :title="item.title"
+                @click="props.onClick"
+                >
+              </v-list-item>
+              <button v-if="item.title != ownerOnLoad" @click="removeOwner(item.title)" class="mx-2">
+                <v-icon class="my-auto" color="pink-accent-4">
+                  {{ mdiCloseCircle }}
+                </v-icon>
+              </button>
+            </div>
```
Comment (MetRonnie) [inline src/components/cylc/Header.vue:79]: Can simplify this. Here I'm using the `append` slot of the `v-list-item` so Vuetify takes care of the positioning.
```suggestion
            <v-list-item
              :title="item.title"
              v-bind="props"
            >
              <template v-slot:append v-if="item.title !== ownerOnLoad">
                <v-icon
                  @click.stop="removeOwner(item.title)"
                  :icon="mdiClose"
                />
              </template>
            </v-list-item>
```

**Comment by markgrahamdawson** [comment]
Probably a bit late now - but for future reference there is a Vue composable for using local storage
https://vueuse.org/core/useLocalStorage/

**Comment by MetRonnie** [comment]
Nice, I'll open an issue for investigating using that. Might be a good training day exercise

File: src/components/cylc/Header.vue [hunk 2]
```diff
@@ -55,8 +56,28 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+                >
+                <template v-slot:append v-if="item.title !== ownerOnLoad">
+                    <v-icon
+                      @click.stop="removeOwner(item.title)"
+                      color="pink-accent-4"
+                      :icon="mdiClose"
+                    />
+                </template>
+              </v-list-item>
+            </div>
```
Comment (MetRonnie) [inline src/components/cylc/Header.vue:77]: The `<div>` here is superfluous
```suggestion
            <v-list-item
              :title="item.title"
              v-bind="props"
            >
              <template v-slot:append v-if="item.title !== ownerOnLoad">
                <v-icon
                  @click.stop="removeOwner(item.title)"
                  color="pink-accent-4"
                  :icon="mdiClose"
                />
              </template>
            </v-list-item>
```

File: src/components/cylc/Header.vue [hunk 3]
```diff
@@ -65,8 +86,27 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+                >
+                <template v-slot:append v-if="item.title !== deploymentOnLoad">
+                    <v-icon
+                      @click.stop="removeDeployment(item.title)"
+                      color="pink-accent-4"
+                      :icon="mdiClose"
+                    />
+                </template>
+              </v-list-item>
+            </div>
```
Comment (MetRonnie) [inline src/components/cylc/Header.vue:107]:
```suggestion
            <v-list-item
              :title="item.title"
              v-bind="props"
            >
              <template v-slot:append v-if="item.title !== deploymentOnLoad">
                <v-icon
                  @click.stop="removeDeployment(item.title)"
                  color="pink-accent-4"
                  :icon="mdiClose"
                />
              </template>
            </v-list-item>
```

File: src/composables/useLocalStorage.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { ref } from 'vue'
+
+export default function useLocalStorage (key, initialValue) {
```
Comment (MetRonnie) [inline src/composables/useLocalStorage.js:3]: Would be good to document this using a JSDoc comment in the format:
```suggestion
/**
 * Description here
 *
 * @param {string} key
 * @param {*} initialValue
 */
export default function useLocalStorage (key, initialValue) {
```
In VSCode if you type `/**` on the line above a function it should come up with an autocomplete suggestion called "JSDoc comment" which will automatically list the params for you

File: tests/unit/composables/useLocalStorage.spec.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  it('should add a value to the local store', () => {
+    addToLocalStorage('test item')
+    expect(Array.from(items.value)).toStrictEqual(['test initial item', 'test item'])
+  })
+
+  it('should remove a value to the local store', () => {
+    removeFromLocalStorage('test item')
+    expect(Array.from(items.value)).toStrictEqual(['test initial item'])
+  })
+})
```
Comment (MetRonnie) [inline tests/unit/composables/useLocalStorage.spec.js:21]: Another thing it might be good to test is that you can't add duplicate entries

File: src/composables/useLocalStorage.js [hunk 6]
```diff
@@ -1,5 +1,18 @@
...
+ * Values are stored in arrays with and can be accessed
+ * via a key which specified by the 'key' parameter.
+ * The local storage array is initialized with a value
+ * specified by the 'initialValue' parameter.
+ * Note: this could be replaced by a VueUse composable in the future
+ * https://vueuse.org/core/useLocalStorage/
+ *
+ * @param {string} key
+ * @param {*} initialValue
+ * @returns {function}
```
Comment (MetRonnie) [inline src/composables/useLocalStorage.js:14]: This is not correct - it returns an object with the fields `itemOnLoad, item, ...` but I wouldn't worry about including that info right now
```suggestion
```

Comment (markgrahamdawson) [reply src/composables/useLocalStorage.js]: Ive added some JSdocs for that composable object

File: src/composables/useLocalStorage.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+ * @name removeFromLocalStorage
+ * @function
+ * @param {String} item Item to remove from local storage
+*/
+
+/**
+ * @typedef {Object} LocalStorage
+ * @property {string} itemOnLoad - The default item on load
+ * @property {ref} item - The selected item
+ * @property {ref} items - The array of items in local storage
```
Comment (MetRonnie) [inline src/composables/useLocalStorage.js:19]: This wasn't quite working with VSCode's type detection features, but I managed to figure out the right syntax.

However, personally I find JSDoc a bit fiddly to work with and if we want this level of type definition we are probably better off adopting TypeScript in future.

```suggestion
 * @callback addToLocalStorage
 * @param {*} item Item to add to local storage
 * @return {void}
*/

/**
 * @callback removeFromLocalStorage
 * @param {*} item Item to remove from local storage
 * @return {void}
*/

/**
 * @typedef {Object} LocalStorage
 * @property {*} itemOnLoad - The default item on load
 * @property {import('vue').Ref} item - The selected item
 * @property {import('vue').Ref<Set>} items - The array of items in local storage
```

**Review by MetRonnie — APPROVED** [review]

**Review by MetRonnie — APPROVED** [review]

**Review by MetRonnie — APPROVED** [review]
Nice

**Review by oliver-sanders — APPROVED** [review]
